// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: namespace/namespace.proto

package namespace

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on CreateRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *CreateRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CreateRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in CreateRequestMultiError, or
// nil if none found.
func (m *CreateRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *CreateRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(m.GetNamespace()) < 1 {
		err := CreateRequestValidationError{
			field:  "Namespace",
			reason: "value length must be at least 1 bytes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	// no validation rules for IgnoreIfExists

	// no validation rules for Description

	if len(errors) > 0 {
		return CreateRequestMultiError(errors)
	}

	return nil
}

// CreateRequestMultiError is an error wrapping multiple validation errors
// returned by CreateRequest.ValidateAll() if the designated constraints
// aren't met.
type CreateRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CreateRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CreateRequestMultiError) AllErrors() []error { return m }

// CreateRequestValidationError is the validation error returned by
// CreateRequest.Validate if the designated constraints aren't met.
type CreateRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreateRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreateRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreateRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreateRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreateRequestValidationError) ErrorName() string { return "CreateRequestValidationError" }

// Error satisfies the builtin error interface
func (e CreateRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreateRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreateRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreateRequestValidationError{}

// Validate checks the field values on CreateResponse with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *CreateResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CreateResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in CreateResponseMultiError,
// or nil if none found.
func (m *CreateResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *CreateResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetItem()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, CreateResponseValidationError{
					field:  "Item",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, CreateResponseValidationError{
					field:  "Item",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetItem()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return CreateResponseValidationError{
				field:  "Item",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for Exists

	if len(errors) > 0 {
		return CreateResponseMultiError(errors)
	}

	return nil
}

// CreateResponseMultiError is an error wrapping multiple validation errors
// returned by CreateResponse.ValidateAll() if the designated constraints
// aren't met.
type CreateResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CreateResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CreateResponseMultiError) AllErrors() []error { return m }

// CreateResponseValidationError is the validation error returned by
// CreateResponse.Validate if the designated constraints aren't met.
type CreateResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreateResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreateResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreateResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreateResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreateResponseValidationError) ErrorName() string { return "CreateResponseValidationError" }

// Error satisfies the builtin error interface
func (e CreateResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreateResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreateResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreateResponseValidationError{}

// Validate checks the field values on ShowRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ShowRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ShowRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ShowRequestMultiError, or
// nil if none found.
func (m *ShowRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ShowRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetId() <= 0 {
		err := ShowRequestValidationError{
			field:  "Id",
			reason: "value must be greater than 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return ShowRequestMultiError(errors)
	}

	return nil
}

// ShowRequestMultiError is an error wrapping multiple validation errors
// returned by ShowRequest.ValidateAll() if the designated constraints aren't met.
type ShowRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ShowRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ShowRequestMultiError) AllErrors() []error { return m }

// ShowRequestValidationError is the validation error returned by
// ShowRequest.Validate if the designated constraints aren't met.
type ShowRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ShowRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ShowRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ShowRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ShowRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ShowRequestValidationError) ErrorName() string { return "ShowRequestValidationError" }

// Error satisfies the builtin error interface
func (e ShowRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sShowRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ShowRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ShowRequestValidationError{}

// Validate checks the field values on ShowResponse with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ShowResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ShowResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ShowResponseMultiError, or
// nil if none found.
func (m *ShowResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ShowResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetItem()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, ShowResponseValidationError{
					field:  "Item",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, ShowResponseValidationError{
					field:  "Item",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetItem()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ShowResponseValidationError{
				field:  "Item",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return ShowResponseMultiError(errors)
	}

	return nil
}

// ShowResponseMultiError is an error wrapping multiple validation errors
// returned by ShowResponse.ValidateAll() if the designated constraints aren't met.
type ShowResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ShowResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ShowResponseMultiError) AllErrors() []error { return m }

// ShowResponseValidationError is the validation error returned by
// ShowResponse.Validate if the designated constraints aren't met.
type ShowResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ShowResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ShowResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ShowResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ShowResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ShowResponseValidationError) ErrorName() string { return "ShowResponseValidationError" }

// Error satisfies the builtin error interface
func (e ShowResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sShowResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ShowResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ShowResponseValidationError{}

// Validate checks the field values on IsExistsRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *IsExistsRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on IsExistsRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// IsExistsRequestMultiError, or nil if none found.
func (m *IsExistsRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *IsExistsRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(m.GetName()) < 1 {
		err := IsExistsRequestValidationError{
			field:  "Name",
			reason: "value length must be at least 1 bytes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return IsExistsRequestMultiError(errors)
	}

	return nil
}

// IsExistsRequestMultiError is an error wrapping multiple validation errors
// returned by IsExistsRequest.ValidateAll() if the designated constraints
// aren't met.
type IsExistsRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m IsExistsRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m IsExistsRequestMultiError) AllErrors() []error { return m }

// IsExistsRequestValidationError is the validation error returned by
// IsExistsRequest.Validate if the designated constraints aren't met.
type IsExistsRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e IsExistsRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e IsExistsRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e IsExistsRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e IsExistsRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e IsExistsRequestValidationError) ErrorName() string { return "IsExistsRequestValidationError" }

// Error satisfies the builtin error interface
func (e IsExistsRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sIsExistsRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = IsExistsRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = IsExistsRequestValidationError{}

// Validate checks the field values on IsExistsResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *IsExistsResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on IsExistsResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// IsExistsResponseMultiError, or nil if none found.
func (m *IsExistsResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *IsExistsResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for Exists

	if len(errors) > 0 {
		return IsExistsResponseMultiError(errors)
	}

	return nil
}

// IsExistsResponseMultiError is an error wrapping multiple validation errors
// returned by IsExistsResponse.ValidateAll() if the designated constraints
// aren't met.
type IsExistsResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m IsExistsResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m IsExistsResponseMultiError) AllErrors() []error { return m }

// IsExistsResponseValidationError is the validation error returned by
// IsExistsResponse.Validate if the designated constraints aren't met.
type IsExistsResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e IsExistsResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e IsExistsResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e IsExistsResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e IsExistsResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e IsExistsResponseValidationError) ErrorName() string { return "IsExistsResponseValidationError" }

// Error satisfies the builtin error interface
func (e IsExistsResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sIsExistsResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = IsExistsResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = IsExistsResponseValidationError{}

// Validate checks the field values on DeleteRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *DeleteRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeleteRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in DeleteRequestMultiError, or
// nil if none found.
func (m *DeleteRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *DeleteRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetId() <= 0 {
		err := DeleteRequestValidationError{
			field:  "Id",
			reason: "value must be greater than 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return DeleteRequestMultiError(errors)
	}

	return nil
}

// DeleteRequestMultiError is an error wrapping multiple validation errors
// returned by DeleteRequest.ValidateAll() if the designated constraints
// aren't met.
type DeleteRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeleteRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeleteRequestMultiError) AllErrors() []error { return m }

// DeleteRequestValidationError is the validation error returned by
// DeleteRequest.Validate if the designated constraints aren't met.
type DeleteRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeleteRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeleteRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeleteRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeleteRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeleteRequestValidationError) ErrorName() string { return "DeleteRequestValidationError" }

// Error satisfies the builtin error interface
func (e DeleteRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeleteRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeleteRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeleteRequestValidationError{}

// Validate checks the field values on DeleteResponse with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *DeleteResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeleteResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in DeleteResponseMultiError,
// or nil if none found.
func (m *DeleteResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *DeleteResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return DeleteResponseMultiError(errors)
	}

	return nil
}

// DeleteResponseMultiError is an error wrapping multiple validation errors
// returned by DeleteResponse.ValidateAll() if the designated constraints
// aren't met.
type DeleteResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeleteResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeleteResponseMultiError) AllErrors() []error { return m }

// DeleteResponseValidationError is the validation error returned by
// DeleteResponse.Validate if the designated constraints aren't met.
type DeleteResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeleteResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeleteResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeleteResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeleteResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeleteResponseValidationError) ErrorName() string { return "DeleteResponseValidationError" }

// Error satisfies the builtin error interface
func (e DeleteResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeleteResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeleteResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeleteResponseValidationError{}

// Validate checks the field values on FavoriteRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *FavoriteRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on FavoriteRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// FavoriteRequestMultiError, or nil if none found.
func (m *FavoriteRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *FavoriteRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetId() <= 0 {
		err := FavoriteRequestValidationError{
			field:  "Id",
			reason: "value must be greater than 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	// no validation rules for Favorite

	if len(errors) > 0 {
		return FavoriteRequestMultiError(errors)
	}

	return nil
}

// FavoriteRequestMultiError is an error wrapping multiple validation errors
// returned by FavoriteRequest.ValidateAll() if the designated constraints
// aren't met.
type FavoriteRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m FavoriteRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m FavoriteRequestMultiError) AllErrors() []error { return m }

// FavoriteRequestValidationError is the validation error returned by
// FavoriteRequest.Validate if the designated constraints aren't met.
type FavoriteRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e FavoriteRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e FavoriteRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e FavoriteRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e FavoriteRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e FavoriteRequestValidationError) ErrorName() string { return "FavoriteRequestValidationError" }

// Error satisfies the builtin error interface
func (e FavoriteRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sFavoriteRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = FavoriteRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = FavoriteRequestValidationError{}

// Validate checks the field values on FavoriteResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *FavoriteResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on FavoriteResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// FavoriteResponseMultiError, or nil if none found.
func (m *FavoriteResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *FavoriteResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return FavoriteResponseMultiError(errors)
	}

	return nil
}

// FavoriteResponseMultiError is an error wrapping multiple validation errors
// returned by FavoriteResponse.ValidateAll() if the designated constraints
// aren't met.
type FavoriteResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m FavoriteResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m FavoriteResponseMultiError) AllErrors() []error { return m }

// FavoriteResponseValidationError is the validation error returned by
// FavoriteResponse.Validate if the designated constraints aren't met.
type FavoriteResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e FavoriteResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e FavoriteResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e FavoriteResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e FavoriteResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e FavoriteResponseValidationError) ErrorName() string { return "FavoriteResponseValidationError" }

// Error satisfies the builtin error interface
func (e FavoriteResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sFavoriteResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = FavoriteResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = FavoriteResponseValidationError{}

// Validate checks the field values on UpdateDescRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *UpdateDescRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdateDescRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdateDescRequestMultiError, or nil if none found.
func (m *UpdateDescRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdateDescRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetId() <= 0 {
		err := UpdateDescRequestValidationError{
			field:  "Id",
			reason: "value must be greater than 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	// no validation rules for Desc

	if len(errors) > 0 {
		return UpdateDescRequestMultiError(errors)
	}

	return nil
}

// UpdateDescRequestMultiError is an error wrapping multiple validation errors
// returned by UpdateDescRequest.ValidateAll() if the designated constraints
// aren't met.
type UpdateDescRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdateDescRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdateDescRequestMultiError) AllErrors() []error { return m }

// UpdateDescRequestValidationError is the validation error returned by
// UpdateDescRequest.Validate if the designated constraints aren't met.
type UpdateDescRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdateDescRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdateDescRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdateDescRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdateDescRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdateDescRequestValidationError) ErrorName() string {
	return "UpdateDescRequestValidationError"
}

// Error satisfies the builtin error interface
func (e UpdateDescRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdateDescRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdateDescRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdateDescRequestValidationError{}

// Validate checks the field values on UpdateDescResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *UpdateDescResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdateDescResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdateDescResponseMultiError, or nil if none found.
func (m *UpdateDescResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdateDescResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetItem()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, UpdateDescResponseValidationError{
					field:  "Item",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, UpdateDescResponseValidationError{
					field:  "Item",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetItem()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return UpdateDescResponseValidationError{
				field:  "Item",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return UpdateDescResponseMultiError(errors)
	}

	return nil
}

// UpdateDescResponseMultiError is an error wrapping multiple validation errors
// returned by UpdateDescResponse.ValidateAll() if the designated constraints
// aren't met.
type UpdateDescResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdateDescResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdateDescResponseMultiError) AllErrors() []error { return m }

// UpdateDescResponseValidationError is the validation error returned by
// UpdateDescResponse.Validate if the designated constraints aren't met.
type UpdateDescResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdateDescResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdateDescResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdateDescResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdateDescResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdateDescResponseValidationError) ErrorName() string {
	return "UpdateDescResponseValidationError"
}

// Error satisfies the builtin error interface
func (e UpdateDescResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdateDescResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdateDescResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdateDescResponseValidationError{}

// Validate checks the field values on ListRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ListRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ListRequestMultiError, or
// nil if none found.
func (m *ListRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ListRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Favorite

	if m.Page != nil {
		// no validation rules for Page
	}

	if m.PageSize != nil {
		// no validation rules for PageSize
	}

	if m.Name != nil {
		// no validation rules for Name
	}

	if len(errors) > 0 {
		return ListRequestMultiError(errors)
	}

	return nil
}

// ListRequestMultiError is an error wrapping multiple validation errors
// returned by ListRequest.ValidateAll() if the designated constraints aren't met.
type ListRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListRequestMultiError) AllErrors() []error { return m }

// ListRequestValidationError is the validation error returned by
// ListRequest.Validate if the designated constraints aren't met.
type ListRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListRequestValidationError) ErrorName() string { return "ListRequestValidationError" }

// Error satisfies the builtin error interface
func (e ListRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListRequestValidationError{}

// Validate checks the field values on ListResponse with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ListResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ListResponseMultiError, or
// nil if none found.
func (m *ListResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ListResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetItems() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ListResponseValidationError{
						field:  fmt.Sprintf("Items[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ListResponseValidationError{
						field:  fmt.Sprintf("Items[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ListResponseValidationError{
					field:  fmt.Sprintf("Items[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	// no validation rules for Count

	// no validation rules for Page

	// no validation rules for PageSize

	if len(errors) > 0 {
		return ListResponseMultiError(errors)
	}

	return nil
}

// ListResponseMultiError is an error wrapping multiple validation errors
// returned by ListResponse.ValidateAll() if the designated constraints aren't met.
type ListResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListResponseMultiError) AllErrors() []error { return m }

// ListResponseValidationError is the validation error returned by
// ListResponse.Validate if the designated constraints aren't met.
type ListResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListResponseValidationError) ErrorName() string { return "ListResponseValidationError" }

// Error satisfies the builtin error interface
func (e ListResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListResponseValidationError{}

// Validate checks the field values on UpdatePrivateRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *UpdatePrivateRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdatePrivateRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdatePrivateRequestMultiError, or nil if none found.
func (m *UpdatePrivateRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdatePrivateRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetId() <= 0 {
		err := UpdatePrivateRequestValidationError{
			field:  "Id",
			reason: "value must be greater than 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	// no validation rules for Private

	if len(errors) > 0 {
		return UpdatePrivateRequestMultiError(errors)
	}

	return nil
}

// UpdatePrivateRequestMultiError is an error wrapping multiple validation
// errors returned by UpdatePrivateRequest.ValidateAll() if the designated
// constraints aren't met.
type UpdatePrivateRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdatePrivateRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdatePrivateRequestMultiError) AllErrors() []error { return m }

// UpdatePrivateRequestValidationError is the validation error returned by
// UpdatePrivateRequest.Validate if the designated constraints aren't met.
type UpdatePrivateRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdatePrivateRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdatePrivateRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdatePrivateRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdatePrivateRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdatePrivateRequestValidationError) ErrorName() string {
	return "UpdatePrivateRequestValidationError"
}

// Error satisfies the builtin error interface
func (e UpdatePrivateRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdatePrivateRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdatePrivateRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdatePrivateRequestValidationError{}

// Validate checks the field values on UpdatePrivateResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *UpdatePrivateResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdatePrivateResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdatePrivateResponseMultiError, or nil if none found.
func (m *UpdatePrivateResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdatePrivateResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetItem()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, UpdatePrivateResponseValidationError{
					field:  "Item",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, UpdatePrivateResponseValidationError{
					field:  "Item",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetItem()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return UpdatePrivateResponseValidationError{
				field:  "Item",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return UpdatePrivateResponseMultiError(errors)
	}

	return nil
}

// UpdatePrivateResponseMultiError is an error wrapping multiple validation
// errors returned by UpdatePrivateResponse.ValidateAll() if the designated
// constraints aren't met.
type UpdatePrivateResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdatePrivateResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdatePrivateResponseMultiError) AllErrors() []error { return m }

// UpdatePrivateResponseValidationError is the validation error returned by
// UpdatePrivateResponse.Validate if the designated constraints aren't met.
type UpdatePrivateResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdatePrivateResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdatePrivateResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdatePrivateResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdatePrivateResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdatePrivateResponseValidationError) ErrorName() string {
	return "UpdatePrivateResponseValidationError"
}

// Error satisfies the builtin error interface
func (e UpdatePrivateResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdatePrivateResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdatePrivateResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdatePrivateResponseValidationError{}

// Validate checks the field values on SyncMembersRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *SyncMembersRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SyncMembersRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// SyncMembersRequestMultiError, or nil if none found.
func (m *SyncMembersRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *SyncMembersRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetId() <= 0 {
		err := SyncMembersRequestValidationError{
			field:  "Id",
			reason: "value must be greater than 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	for idx, item := range m.GetEmails() {
		_, _ = idx, item

		if err := m._validateEmail(item); err != nil {
			err = SyncMembersRequestValidationError{
				field:  fmt.Sprintf("Emails[%v]", idx),
				reason: "value must be a valid email address",
				cause:  err,
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

	}

	if len(errors) > 0 {
		return SyncMembersRequestMultiError(errors)
	}

	return nil
}

func (m *SyncMembersRequest) _validateHostname(host string) error {
	s := strings.ToLower(strings.TrimSuffix(host, "."))

	if len(host) > 253 {
		return errors.New("hostname cannot exceed 253 characters")
	}

	for _, part := range strings.Split(s, ".") {
		if l := len(part); l == 0 || l > 63 {
			return errors.New("hostname part must be non-empty and cannot exceed 63 characters")
		}

		if part[0] == '-' {
			return errors.New("hostname parts cannot begin with hyphens")
		}

		if part[len(part)-1] == '-' {
			return errors.New("hostname parts cannot end with hyphens")
		}

		for _, r := range part {
			if (r < 'a' || r > 'z') && (r < '0' || r > '9') && r != '-' {
				return fmt.Errorf("hostname parts can only contain alphanumeric characters or hyphens, got %q", string(r))
			}
		}
	}

	return nil
}

func (m *SyncMembersRequest) _validateEmail(addr string) error {
	a, err := mail.ParseAddress(addr)
	if err != nil {
		return err
	}
	addr = a.Address

	if len(addr) > 254 {
		return errors.New("email addresses cannot exceed 254 characters")
	}

	parts := strings.SplitN(addr, "@", 2)

	if len(parts[0]) > 64 {
		return errors.New("email address local phrase cannot exceed 64 characters")
	}

	return m._validateHostname(parts[1])
}

// SyncMembersRequestMultiError is an error wrapping multiple validation errors
// returned by SyncMembersRequest.ValidateAll() if the designated constraints
// aren't met.
type SyncMembersRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SyncMembersRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SyncMembersRequestMultiError) AllErrors() []error { return m }

// SyncMembersRequestValidationError is the validation error returned by
// SyncMembersRequest.Validate if the designated constraints aren't met.
type SyncMembersRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SyncMembersRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SyncMembersRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SyncMembersRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SyncMembersRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SyncMembersRequestValidationError) ErrorName() string {
	return "SyncMembersRequestValidationError"
}

// Error satisfies the builtin error interface
func (e SyncMembersRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSyncMembersRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SyncMembersRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SyncMembersRequestValidationError{}

// Validate checks the field values on SyncMembersResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *SyncMembersResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SyncMembersResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// SyncMembersResponseMultiError, or nil if none found.
func (m *SyncMembersResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *SyncMembersResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetItem()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, SyncMembersResponseValidationError{
					field:  "Item",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, SyncMembersResponseValidationError{
					field:  "Item",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetItem()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return SyncMembersResponseValidationError{
				field:  "Item",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return SyncMembersResponseMultiError(errors)
	}

	return nil
}

// SyncMembersResponseMultiError is an error wrapping multiple validation
// errors returned by SyncMembersResponse.ValidateAll() if the designated
// constraints aren't met.
type SyncMembersResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SyncMembersResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SyncMembersResponseMultiError) AllErrors() []error { return m }

// SyncMembersResponseValidationError is the validation error returned by
// SyncMembersResponse.Validate if the designated constraints aren't met.
type SyncMembersResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SyncMembersResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SyncMembersResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SyncMembersResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SyncMembersResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SyncMembersResponseValidationError) ErrorName() string {
	return "SyncMembersResponseValidationError"
}

// Error satisfies the builtin error interface
func (e SyncMembersResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSyncMembersResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SyncMembersResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SyncMembersResponseValidationError{}

// Validate checks the field values on TransferRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *TransferRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on TransferRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// TransferRequestMultiError, or nil if none found.
func (m *TransferRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *TransferRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetId() <= 0 {
		err := TransferRequestValidationError{
			field:  "Id",
			reason: "value must be greater than 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(m.GetNewAdminEmail()) < 1 {
		err := TransferRequestValidationError{
			field:  "NewAdminEmail",
			reason: "value length must be at least 1 bytes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if err := m._validateEmail(m.GetNewAdminEmail()); err != nil {
		err = TransferRequestValidationError{
			field:  "NewAdminEmail",
			reason: "value must be a valid email address",
			cause:  err,
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return TransferRequestMultiError(errors)
	}

	return nil
}

func (m *TransferRequest) _validateHostname(host string) error {
	s := strings.ToLower(strings.TrimSuffix(host, "."))

	if len(host) > 253 {
		return errors.New("hostname cannot exceed 253 characters")
	}

	for _, part := range strings.Split(s, ".") {
		if l := len(part); l == 0 || l > 63 {
			return errors.New("hostname part must be non-empty and cannot exceed 63 characters")
		}

		if part[0] == '-' {
			return errors.New("hostname parts cannot begin with hyphens")
		}

		if part[len(part)-1] == '-' {
			return errors.New("hostname parts cannot end with hyphens")
		}

		for _, r := range part {
			if (r < 'a' || r > 'z') && (r < '0' || r > '9') && r != '-' {
				return fmt.Errorf("hostname parts can only contain alphanumeric characters or hyphens, got %q", string(r))
			}
		}
	}

	return nil
}

func (m *TransferRequest) _validateEmail(addr string) error {
	a, err := mail.ParseAddress(addr)
	if err != nil {
		return err
	}
	addr = a.Address

	if len(addr) > 254 {
		return errors.New("email addresses cannot exceed 254 characters")
	}

	parts := strings.SplitN(addr, "@", 2)

	if len(parts[0]) > 64 {
		return errors.New("email address local phrase cannot exceed 64 characters")
	}

	return m._validateHostname(parts[1])
}

// TransferRequestMultiError is an error wrapping multiple validation errors
// returned by TransferRequest.ValidateAll() if the designated constraints
// aren't met.
type TransferRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m TransferRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m TransferRequestMultiError) AllErrors() []error { return m }

// TransferRequestValidationError is the validation error returned by
// TransferRequest.Validate if the designated constraints aren't met.
type TransferRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e TransferRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e TransferRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e TransferRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e TransferRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e TransferRequestValidationError) ErrorName() string { return "TransferRequestValidationError" }

// Error satisfies the builtin error interface
func (e TransferRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sTransferRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = TransferRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = TransferRequestValidationError{}

// Validate checks the field values on TransferResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *TransferResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on TransferResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// TransferResponseMultiError, or nil if none found.
func (m *TransferResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *TransferResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetItem()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, TransferResponseValidationError{
					field:  "Item",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, TransferResponseValidationError{
					field:  "Item",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetItem()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return TransferResponseValidationError{
				field:  "Item",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return TransferResponseMultiError(errors)
	}

	return nil
}

// TransferResponseMultiError is an error wrapping multiple validation errors
// returned by TransferResponse.ValidateAll() if the designated constraints
// aren't met.
type TransferResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m TransferResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m TransferResponseMultiError) AllErrors() []error { return m }

// TransferResponseValidationError is the validation error returned by
// TransferResponse.Validate if the designated constraints aren't met.
type TransferResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e TransferResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e TransferResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e TransferResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e TransferResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e TransferResponseValidationError) ErrorName() string { return "TransferResponseValidationError" }

// Error satisfies the builtin error interface
func (e TransferResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sTransferResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = TransferResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = TransferResponseValidationError{}
